receiver/prometheusexecreceiver/end_to_end_test.go

```diff
@@ -0,0 +1,130 @@
+func TestEndToEnd(t *testing.T) {
+	// Load the config from the yaml file
+	factories, err := componenttest.ExampleComponents()
+	assert.NoError(t, err)
+
+	factory := &Factory{}
+	factories.Receivers[factory.Type()] = factory
+
+	config, err := configtest.LoadConfigFile(t, path.Join(".", "testdata", "config.yaml"), factories)
+
+	assert.NoError(t, err)
+	assert.NotNil(t, config)
+
+	// Receiver without exec key, error expected and checked within function
+	execErrorTest(t, config.Receivers["prometheus_exec"])
+
+	// Normal test with port defined, expose metrics from fake exporter and make sure they're scraped/received
+	//endToEndScrapeTest(t, config.Receivers["prometheus_exec/end_to_end_test/1"], "end-to-end port defined")
+	// 2do: fix test
+
+	// Normal test with port undefined by user, same as previous test
+	endToEndScrapeTest(t, config.Receivers["prometheus_exec/end_to_end_test/2"], "end-to-end port not defined")
+}
+
+// execErrorTest makes sure the config passed throws an error, since it's missing the exec key
+func execErrorTest(t *testing.T, errorReceiverConfig configmodels.Receiver) {
+	wrapper := new(zap.NewNop(), errorReceiverConfig.(*Config), nil)
+
+	err := wrapper.Start(context.Background(), componenttest.NewNopHost())
+	if err == nil {
+		t.Errorf("end_to_end_test.go didn't get error, was expecting one since this config has no 'exec' key")
+	}
+}
+
+// endToEndScrapeTest scrapes a test endpoint (test_prometheus_exporter.go) twice, and between each scrape yields the execution with Sleep() to wait for the subprocess (exporter) to restar
+// - wait time is about 1s - to fail and restart, meaning it verifies three things: the scrape is successful (twice), the process was restarted correctly when failed and the underlying
+// Prometheus receiver was correctly stopped and then restarted. For extra testing the metrics values are different every time the subprocess exporter is started
+// And the uniqueness of the metric scraped is verified
+func endToEndScrapeTest(t *testing.T, receiverConfig configmodels.Receiver, testName string) {
+	sink := &exportertest.SinkMetricsExporterOld{}
+	wrapper := new(zap.NewNop(), receiverConfig.(*Config), sink)
+
+	ctx := context.Background()
+	err := wrapper.Start(ctx, componenttest.NewNopHost())
+	if err != nil {
+		t.Errorf("end_to_end_test.go got error = %w", err)
+	}
+	defer wrapper.Shutdown(ctx)
+
+	var metrics []consumerdata.MetricsData
+
+	// Make sure a first scrape works by checking for metrics in the test metrics exporter "sink", only return true when there are metrics
+	const waitFor = 15 * time.Second
+	const tick = 500 * time.Millisecond
+	require.Eventuallyf(t, func() bool {
+		got := sink.AllMetrics()
+		return len(got) != 0
+	}, waitFor, tick, "No metrics were collected after %v for the first scrape (%v)", waitFor, testName)
+
+	// Wait for subprocess to restart - wait time is about 1s - and allow the other test to run in parallel
+	time.Sleep(1000 * time.Millisecond)
```

Comment:
Long waits like this are bad for two reasons:

1. Occasionally things may take longer than expected (e.g. if test machine is under heavy load) causing the test to fail / be flaky
2. This causes a long wait. You generally want tests to be as fast as possible otherwise build times can easily get out of hand.

A better way of managing this would be to have a channel that your tests can subscribe to that is notified when the status changes (or at least when the process is finished restarting)

That might be difficult for you to do here though. Can you at least avoid this `time.Sleep` here though?